### ISSUE TITLE
BatchNorm Behavior for is_train=false

### ISSUE THREAD

**Issue by lukemetz** [dmlc/mxnet#960]
Currently there is no support for running a backward pass in is_train = False. This makes stuff like (http://arxiv.org/abs/1312.6034) and #895 impossible to do.

Can this be changed to doing the true derivative with respect to the is_train parameter from the forward pass? As of now there is a hardcoded is_train=true passed in (https://github.com/dmlc/mxnet/blob/6cc09e6d3716089ffbf050295d596e85cf6b809d/src/symbol/graph_executor.cc#L842)

Currently the behaviour is rather confusing, when trying to take a derivative through a batch norm layer the gradients returned are all nans.

I have the changes locally to do this locally (for a few ops at least) but I wanted to get some feedback before I open a PR.

**Comment by tqchen** [comment]
Yes, I think we can do that. For BatchNorm specificially. Note that gradient of BN need to be changed, as the formula becomes
```
out = (in - moving_mean)/ moving_var * beta + gamma
```
and moving mean and ```moving_var```  are constant now.

Loop in @mavenlin and @antinucleon 

Please propose a proper backprop rule and we can check and then do a PR

**Comment by lukemetz** [comment]
I would propose having the `is_train` flag saved inside the graph executor as a member variable of GraphExecutor when forward is called (and partialForward) and that saved value used / passed into RunOps on the backward pass.

I am normally hesitant of this hidden data passing, but my understanding of how mxnet deals with data needed for ops is that one needs to always call a forward pass before calling a backward pass to calculate needed data. Thus having a possible mismatch between the two (one to forward and one to backward) could cause user error. Another option would be to shove this state up to the front end language and keep everything very explicit on the cpp side. I am not sure which is preferable.

With this in place, ops that use the is_train flag on the forward pass would be changed to correctly use this flag.
These currently just include:
* [ ] batchnorm
* [ ] dropout
* [ ] (others?)

How does this sound?

**Comment by piiswrong** [comment]
@tqchen Is exec.backward() valid when is_train=False? I thought the internal activations will all be discarded?

**Comment by lukemetz** [comment]
@piiswrong I did a quick check of this initially but this appears not to be the case. In addition, I was able to validate that the gradients back through some simple tests, using numeric diff, were correct.

I for sure didn't look everywhere so it would be good to get a double check.
